src/System.Private.CoreLib/shared/System/IO/Stream.cs

```diff
@@ -189,6 +189,80 @@ private int GetCopyBufferSize()
             return bufferSize;
         }
 
+        public virtual void CopyTo(ReadOnlySpanAction<byte, object?> callback, object? state, int bufferSize)
+        {
+            if (callback == null) throw new ArgumentNullException(nameof(callback));
+
+            CopyTo(new WriteCallbackStream(callback, state), bufferSize);
+        }
+
+        public virtual Task CopyToAsync(Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask> callback, object? state, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (callback == null) throw new ArgumentNullException(nameof(callback));
+
+            return CopyToAsync(new WriteCallbackStream(callback, state), bufferSize, cancellationToken);
+        }
+
+        private sealed class WriteCallbackStream : Stream
+        {
+            private readonly ReadOnlySpanAction<byte, object?>? _action;
+            private readonly Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask>? _func;
+            private readonly object? _state;
+
+            public WriteCallbackStream(ReadOnlySpanAction<byte, object?> action, object? state)
+            {
+                _action = action;
+                _state = state;
+            }
+
+            public WriteCallbackStream(Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask> func, object? state)
+            {
+                _func = func;
+                _state = state;
+            }
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                Write(new ReadOnlySpan<byte>(buffer, offset, count));
+            }
+
+            public override void Write(ReadOnlySpan<byte> span)
+            {
+                if (_action != null)
+                    _action(span, _state);
+                else if (_func != null)
+                    _func(span.ToArray(), _state, CancellationToken.None).GetAwaiter().GetResult();
+                else
+                    throw new NotSupportedException();
+            }
+
+            public override Task WriteAsync(byte[] buffer, int offset, int length, CancellationToken cancellationToken)
+            {
+                return WriteAsync(new ReadOnlyMemory<byte>(buffer, offset, length), cancellationToken).AsTask();
+            }
+
+            public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+            {
+                if (_func != null)
+                    await _func(buffer, _state, cancellationToken);
+                else if (_action != null)
+                    _action(buffer.Span, _state);
+                else
+                    throw new NotSupportedException();
+            }
```

Comment:
Please make this:
```C#
public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
{
    if (_func != null)
    {
        return _func(buffer, _state, cancellationToken);
    }

    // In case a poorly implemented CopyTo(Stream, ...) method decides to call
    // the destination stream's WriteAsync rather than Write, we make it work,
    // but this does not need to be efficient.
    Debug.Assert(_action != null);
    try
    {
        cancellationToken.ThrowIfCancellationRequested();
        _action(buffer.Span, _state);
        return default;
    }
    catch (Exception e)
    {
        return new ValueTask(Task.FromException(e));
    }
}
```
Note this is not an `async` method, we don't need a fallback to handle the impossible case of both delegates being null, we handle the case where the _action may throw by wrapping it in a try/catch and marshaling the exception out via the returned task, and we check up front for cancellation in the _action case where it's otherwise ignored.